### DISCUSSION TITLE
Error: Unable to locate executable file: 'dotnet'

### DISCUSSION THREAD

**Discussion by RobBowman** [actions/virtual-environments#8014] [General questions] [answered]
Hi, I'm still trying to figure this out. This is what I've done so far:

- Ran GenerateResourcesAndImage with an -ImageType of "Ubuntu2204". This seems to have completed successfully
- Ran "az disk create" followed by "az image create"
- Created a vm scale set pointing at the new image
- Created an azure vm scale set pointing at the new image
- Created a new devops agent pool that uses the vm scale set
- Ran a devops pipeline using the new agent pool

On running the pipeline, I got the following error: _Error: Unable to locate executable file: 'dotnet'_

Can anyone see where I've gone wrong? What would be the troubleshooting steps to try and resolve?

**Comment by ilia-shipitsin** [comment]
assuming your are trying to build an image identical to github one, you can compare output of "which dotnet" and "dpkg -l | grep dotnet" on your image and github.

https://github.com/ilia-shipitsin/ubuntu-dotnet/actions/runs/5728244156/job/15522354381

**Reply by RobBowman** [reply]
@ilia-shipitsin thanks for your answer but I'm not yet able to connect to the vm instance. I've not used ssh before but [this](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/ssh-from-windows#create-a-vm-using-your-key) suggests a public key needs to be provided when the vm is created?

**Reply by RobBowman** [reply]
Progress... I have now managed to connect via ssh with username and password :)

**Reply by RobBowman** [reply]
After connecting to my VM scale set instance via SSH, I ran "which dotnet" - that gave no response. I then ran "dpkg -l | grep dotnet" and that gave "--info needs a .deb filename argument"

**Reply by RobBowman** [reply]
I ran _sudo find / -type f -name "dotnet.exe"_ but it returned no results

**Reply by ilia-shipitsin** [reply]
not "dotnet.exe" on Linux it is named "dotnet".

it is not a folder in /usr/bin, just a binary (executable) file

no response from 'which" means either /usr/bin is not in PATH variable (run "echo $PATH" to check) or package is not installed

**Comment by RobBowman** [comment]
I guess there should be a "dotnet" folder in usr/bin? Mine seems to be missing

![image](https://github.com/actions/runner-images/assets/6296919/c016eb83-6bf3-4530-8c00-fb8eb4082e15)

**Reply by vpolikarpov-akvelon** [reply, marked as answer]
Hey @RobBowman. You are right, a symlink to `dotnet` executable should be in `/usr/bin/`. As you don't see it then dotnet haven't been installed probably.

But in general this doesn't look like screenshot taken from Ubuntu2204 image. I tried to list this directory on real runner in GitHub and there are a lot more stuff inside (see image below). I assume that either you have built UbuntuMinimal image instead or generation process haven't succeeded. Could you please check these once more?

<details><summary>Screenshot</summary>

![image](https://github.com/actions/runner-images/assets/126792224/eb6702bc-f0f5-4ebd-94fd-74f2e4011188)
</details>

**Reply by RobBowman** [reply]
@vpolikarpov-akvelon thanks for your answer. The line I ran to create the image was: 
```
GenerateResourcesAndImage -SubscriptionId $sourceSubId -ResourceGroupName $sourceRG -AzureLocation "UK South" -ImageType "Ubuntu2204" .
```
 I'm going to try and create create a new devops pipeline that will run the creation of the vm image using [this](https://github.com/YannickRe/azuredevops-buildagents) repo. I'll report back on how that goes.

**Reply by RobBowman** [reply]
So, I got everything setup using Yannick's pipeline. The image created contained everything I needed to build the dotnet app. So, I guess previously I somehow managed to build a base image. 

For info, I have blogged about the experience [here](https://robbowman.github.io/2023/08/03/vmss-hosted-devops-agent.html). 

Thanks for helping out and supporting this vital repo!
